etl/steps/data/garden/pew/2024-06-03/same_sex_marriage.py

```diff
@@ -0,0 +1,196 @@
+REGIONS = ["Europe", "Asia", "North America", "South America", "Africa", "Oceania", "World"]
+
+# Define fraction of allowed NaNs per year
+FRAC_ALLOWED_NANS_PER_YEAR = 0.2
+
+
+def run(dest_dir: str) -> None:
+    #
+    # Load inputs.
+    #
+    # Load meadow dataset, regions and population
+    ds_meadow = paths.load_dataset("same_sex_marriage")
+    ds_regions = paths.load_dataset("regions")
+    ds_population = paths.load_dataset("population")
+
+    # Read table from meadow dataset.
+    tb = ds_meadow["same_sex_marriage"].reset_index()
+
+    #
+    # Process data.
+    #
+    tb = geo.harmonize_countries(
+        df=tb,
+        countries_file=paths.country_mapping_path,
+    )
+
+    tb = explode_country_years(tb=tb)
+
+    tb = add_country_counts_and_population_by_status(
+        tb=tb,
+        columns=["legal_status"],
+        ds_regions=ds_regions,
+        ds_population=ds_population,
+        regions=REGIONS,
+        missing_data_on_columns=False,
+    )
+
+    # Redefine legal_status_not_legal_pop as legal_status_not_legal_pop + legal_status_missing_pop
+    tb["legal_status_Not legal_pop"] = tb["legal_status_Not legal_pop"] + tb["legal_status_missing_pop"]
+
+    # Drop legal_status_missing_pop and legal_status_not_legal_count
+    tb = tb.drop(columns=["legal_status_missing_pop", "legal_status_Not legal_count"])
+
+    tb = tb.format(["country", "year"])
+
+    #
+    # Save outputs.
+    #
+    # Create a new garden dataset with the same metadata as the meadow dataset.
+    ds_garden = create_dataset(
+        dest_dir, tables=[tb], check_variables_metadata=True, default_metadata=ds_meadow.metadata
+    )
+
+    # Save changes in the new garden dataset.
+    ds_garden.save()
+
+
+def explode_country_years(tb: Table) -> Table:
+    """
+    Create a row for each country-year from the minimum year to LATEST_YEAR
+    """
+
+    # Add category to define legal status
+    tb["legal_status"] = "Legal"
+    tb["legal_status"] = tb["legal_status"].copy_metadata(tb["country"])
+
+    min_year = tb["year"].min() - 1
+
+    # Define the list of countries
+    countries = tb["country"].unique().tolist()
+
+    # Create a table of years and another of countries
+    tb_years = Table({"year": range(min_year, LATEST_YEAR + 1)})
+    tb_countries = Table({"country": countries})
+
+    # Create a new table with the cartesian product of tb_years and tb_countries
+    tb_countries_years = pr.merge(tb_years, tb_countries, how="cross")
+
+    # Merge this table with tb
+    tb = pr.merge(tb_countries_years, tb, how="left", on=["country", "year"], short_name=paths.short_name)
+
+    # Fill data forward after the first incidence of "Legal" in each country
+    tb["legal_status"] = tb.groupby("country")["legal_status"].ffill()
+
+    # Add "Not legal" for missing legal status
+    tb.loc[tb["legal_status"].isnull(), "legal_status"] = "Not legal"
+
+    return tb
+
+
+def add_country_counts_and_population_by_status(
```

Comment:
It's a little hard to follow what's going on in here. Perhaps it might be better as a few smaller functions. Also would be good to add some sense checks/asserts at a few critical points.

Reply:
Thanks Fiona. I am thinking I can adapt this function and publish it on `geo`, because it has been very useful to construct aggregates from the categories of categorical variables. So I will hold changes here to adapt it more generally and publish it for the rest.